rfc/image-discoverability.md

```diff
@@ -0,0 +1,127 @@
+
+```
+s3://<BUCKET-NAME>/images/<IMAGE-VERSION-UID>.json
+```
+
+```json
+{
+  "aws": {
+    "us-east-1": "ami-123",
+    "us-west-2": "ami-456",
+    "eu-west-1": "ami-789"
+  },
+  "azure": {
+    "cvm": "azure-cvm-123",
+    "trustedlaunch": "azure-trusted-launch-123"
+  },
+  "gcp": "gcp-image-123",
+  "qemu": "https://cdn.edgeless.systems/.../qemu.raw"
+}
+```
+
+- For AWS, the image lookup table contains the AMI IDs for each region
+- For Azure, the image lookup table contains the image IDs for the CVM and Trusted Launch images
+- For GCP, the image lookup table contains the image ID
+- For QEMU, the image lookup table contains a URL to the QEMU image
+
+This document is not signed and can be extended in the future to include more image references (e.g. if an image is replicated to a new AWS region).
+The same document can be used to identify old images that are no longer used and can be deleted for cost optimization.
+
+## Image measurements
+
+This RFC is not about the image measurements themselves, but about how they are stored and looked up.
+The format of the image measurements is described in the [secure software distribution RFC](secure-software-distribution.md).
+
+The image measurements are stored in a folder structure in S3 that is organized by CSP and `image version uid`.
+
+```
+s3://<BUCKET-NAME>/measurements/<CSP>/<IMAGE-VERSION-UID>/measurements.yaml
+s3://<BUCKET-NAME>/measurements/<CSP>/<IMAGE-VERSION-UID>/measurements.yaml.sig
+```
+
+## CLI image discovery
+
+The CLI needs to be able to discover the image references for a given `image version uid`.
+By default, the CLI will prefill the `image` field of the `constellation-conf.yaml` when `constellation config generate <CSP>` is run with a hardcoded `image version uid` (e.g. `v2.2.0-cc0de5c68d41f31dd0b284d574f137e0b0ad106b`).
+The `image` field is independent of the CSP and is a used to discover the CSP-specific image reference as needed for the following operations:
+
+- `constellation create`
+- `constellation upgrade apply`
+
+The CLI can find a CSP- and region specific image reference by looking up the `image version uid` in the following order:
+
+- if a local file `<IMAGE-VERSION-UID>.json` exists, use the lookup table in that file
+- otherwise, load the image lookup table from a well known URL (e.g. `https://cdn.confidential.cloud/images/<IMAGE-VERSION-UID>.json`) and use the lookup table in that file
+- choose the CSP-specific image reference for the current region and security type:
+  - On AWS, use the AMI ID for the current region (e.g. `.aws.us-east-1`)
+  - On Azure, use the image ID for the security type (CVM or Trusted Launch) (e.g. `.azure.cvm`)
+  - On GCP, use the only image ID (e.g. `.gcp`)
+
+This allows customers to upload images to their own cloud subscription and use them with the CLI by providing the image lookup table as a local file.
```

Comment:
Is the Trusted Launch workflow supposed to be
* download the image
* upload the image to my cloud subscription
* create a local lookup table file

?

Reply:
I see three possible ways to expose this functionality to our users:

1) Our scripts provided to customers that they can use to reupload images also create the local lookup table file. User places the file in the workspace.
2) A larger org may want to have their own "repository". A HTTP server that hosts their lookup tables with their images and uses the same paths as described in this document. Only the base URL would be their own. They could then set the repository in the `constellation-conf.yaml`. I could extend the RFC to explicitly support this.
3) We add an alternative field or a special scheme to be used in the exiting version field: `imageReference` could override the `image` version field with a literal cloud provider image reference OR `gcp://<image-reference>` if the same field should be reused.

1) is the easiest to implement and allows users as well as developers to make local overrides while still using the same versions as the official images. This allows our tools to still reason about what a version is compatible with -- even if the image reference points to a copy of the image.
2) Is only feasible for larger orgs. It would be the most elegant solution in my opinion.
3) Has a large drawback: We cannot reason about the version of the provided image reference anymore. If we want to implement any form of compatibility checks, not knowing what image version a reference belongs to makes this impossible. Even if we use two fields (version and reference) individually, a user could update one field and we would still make wrong decisions based on that.

Reply:
@thomasten any preferences on the preferred implementation of this? If there are none I would leave the RFC as-is and add the customer repository as a possible future extension.

Reply:
sounds good